executor/aggregate.go

```diff
@@ -395,32 +396,54 @@ func (w *HashAggPartialWorker) shuffleIntermData(sc *stmtctx.StatementContext, f
 }
 
 // getGroupKey evaluates the group items and args of aggregate functions.
-func (w *HashAggPartialWorker) getGroupKey(sc *stmtctx.StatementContext, row chunk.Row) ([]byte, error) {
-	w.groupValDatums = w.groupValDatums[:0]
-	for _, item := range w.groupByItems {
-		v, err := item.Eval(row)
+func getGroupKey(ctx sessionctx.Context, input *chunk.Chunk, groupKey [][]byte, groupByItems []expression.Expression) ([][]byte, error) {
+	numRows := input.NumRows()
+	if len(groupKey) < numRows {
+		groupKeyLen := len(groupKey)
+		for i := 0; i < groupKeyLen; i++ {
+			groupKey[i] = groupKey[i][:0]
+		}
+		for i := groupKeyLen; i < numRows; i++ {
+			groupKey = append(groupKey, make([]byte, 0, 10*len(groupByItems)))
+		}
+	} else {
+		for i := 0; i < numRows; i++ {
+			groupKey[i] = groupKey[i][:0]
+		}
+	}
+
+	for _, item := range groupByItems {
+		tp := item.GetType()
+		buf, err := expression.GetColumn(tp.EvalType(), numRows)
 		if err != nil {
 			return nil, err
 		}
-		// This check is used to avoid error during the execution of `EncodeDecimal`.
-		if item.GetType().Tp == mysql.TypeNewDecimal {
-			v.SetLength(0)
+
+		if groupKey, err = expression.VectorizedGetGroupKey(ctx, groupKey, item, tp, input, buf); err != nil {
+			return nil, err
 		}
-		w.groupValDatums = append(w.groupValDatums, v)
+		expression.PutColumn(buf)
 	}
-	var err error
-	w.groupKey, err = codec.EncodeValue(sc, w.groupKey[:0], w.groupValDatums...)
-	return w.groupKey, err
+	return groupKey, nil
 }
 
-func (w baseHashAggWorker) getPartialResult(sc *stmtctx.StatementContext, groupKey []byte, mapper aggPartialResultMapper) []aggfuncs.PartialResult {
-	partialResults, ok := mapper[string(groupKey)]
-	if !ok {
-		partialResults = make([]aggfuncs.PartialResult, 0, len(w.aggFuncs))
+func (w baseHashAggWorker) getPartialResult(sc *stmtctx.StatementContext, groupKey [][]byte, mapper aggPartialResultMapper) [][]aggfuncs.PartialResult {
+	numRows := len(groupKey)
+	partialResults := make([][]aggfuncs.PartialResult, numRows)
+	for i := 0; i < numRows; i++ {
+		partialResults[i] = make([]aggfuncs.PartialResult, 0, len(w.aggFuncs))
+	}
+	for i := 0; i < numRows; i++ {
+		var ok bool
+		partialResults[i], ok = mapper[string(hack.String(groupKey[i]))]
```

Comment:
Actually we can extract `string(hack.String(groupKey[i]))` as an individual variable, thus we can reuse it in line 446.